### PR TITLE
[wip] add appdynamics agent to cluster

### PR DESCRIPTION
**Pull request by linki**
adds a daemonset running the logwatcher and appdynamics machine agent. you have to provide the `appdynamics` key on the command line for create and update.

this is more of a draft to get something running. todo:
* make appd account configurable
* put secrets in a secret (although it won't be more secret because of it)

i do not like providing the appd key on each operation but in the past we decided not to introduce `if/else` templating in our cloud config. i don't see another way currently.

### CODE REVIEW

File: cluster/userdata-master.yaml [hunk 1]
```diff
@@ -906,6 +906,78 @@ write_files:
...
+            labels:
+              app: appdynamics
+              component: logging
+            annotations:
+              scheduler.alpha.kubernetes.io/critical-pod: ''
+              scheduler.alpha.kubernetes.io/tolerations: '[{"key":"CriticalAddonsOnly", "operator":"Exists"}]'
+          spec:
+            containers:
+            - name: log-watcher
+              image: registry.opensource.zalan.do/eagleeye/k8s-log-watcher:0.5-b2
```
Comment (mohabusama) [inline cluster/userdata-master.yaml]: 0.7 is the latest

File: cluster/userdata-master.yaml [hunk 2]
```diff
@@ -906,6 +906,78 @@ write_files:
...
+              - name: CLUSTER_NODE_NAME
+                valueFrom:
+                  fieldRef:
+                    fieldPath: spec.nodeName
+              - name: WATCHER_KUBE_URL
+                value: http://127.0.0.1:8001
+              - name: WATCHER_DEBUG
+                value: "1"
+            - name: appdynamics-agent
+              image: pierone.stups.zalan.do/monitoring-appdynamics/appdynamics-machine-agent:0.2-b2
```
Comment (mohabusama) [inline cluster/userdata-master.yaml]: 0.3 is the latest

File: cluster/cluster.py [hunk 3]
```diff
@@ -307,20 +309,24 @@ def same_user_data(enc1, enc2):
...
+def update(stack_name, version,  dry_run, force, instance_type, master_nodes, worker_nodes, postpone, max_worker_nodes, appdynamics_access_key):
     '''
     Update Kubernetes cluster
     '''
     existing_user_data_master = get_launch_configuration_user_data(stack_name, version, 'Master')
     existing_user_data_worker = get_launch_configuration_user_data(stack_name, version, 'Worker')
     worker_shared_secret = get_worker_shared_secret(existing_user_data_worker)
-    variables = get_cluster_variables(stack_name, version, worker_shared_secret)
+    variables = get_cluster_variables(stack_name, version, appdynamics_access_key, worker_shared_secret)
+    if dry_run:
```
Comment (mohabusama) [inline cluster/cluster.py]: May be a dry-run would print user_data_master and worker as well?

File: cluster/userdata-master.yaml (hunk 1 above)
Comment (mohabusama) [inline cluster/userdata-master.yaml]: Actually ``0.8`` with multiple namspaces support

**Comment by hjacobs** [comment]
What about worker nodes? We want to streamm app logs too, right?

**Comment by hjacobs** [comment]
:+1:

**Comment by mikkeloscar** [comment]
:+1:

File: cluster/cluster.py (hunk 3 above)
Comment (mikkeloscar) [inline cluster/cluster.py]: Opened a separate issue for this: #123

Comment (linki) [inline cluster/cluster.py]: good point, thanks!

**Pull request merged**